File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/database/witnessing/WitnessDao.java
package com.github.dedis.popstellar.repository.database.witnessing;

import androidx.room.*;

import com.github.dedis.popstellar.model.objects.security.PublicKey;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;

@Dao
public interface WitnessDao {

  @Insert(onConflict = OnConflictStrategy.REPLACE)
  Completable insert(List<WitnessEntity> witnessEntities);

  @Query("SELECT witness FROM witnesses WHERE lao_id = :laoId")
  Single<List<PublicKey>> getWitnessesByLao(String laoId);

  @Query("SELECT COUNT(*) FROM witnesses WHERE lao_id = :laoId AND witness = :witness")
  int isWitness(String laoId, PublicKey witness);
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/database/witnessing/WitnessDao.kt
package com.github.dedis.popstellar.repository.database.witnessing

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.github.dedis.popstellar.model.objects.security.PublicKey
import io.reactivex.Completable
import io.reactivex.Single

@Dao
interface WitnessDao {
  @Insert(onConflict = OnConflictStrategy.REPLACE)
  fun insertAll(witnessEntities: List<WitnessEntity>): Completable

  @Query("SELECT witness FROM witnesses WHERE lao_id = :laoId")
  fun getWitnessesByLao(laoId: String): Single<List<PublicKey>>

  @Query("SELECT COUNT(*) FROM witnesses WHERE lao_id = :laoId AND witness = :witness")
  fun isWitness(laoId: String, witness: PublicKey): Int
}
